enzyme/Enzyme/MLIR/Analysis/Lattice.h

```diff
@@ -0,0 +1,364 @@
+
+    ChangeResult result = ChangeResult::NoChange;
+    for (DistinctAttr key : keys) {
+      auto lhsIt = map.find(key);
+      auto rhsIt = rhs.map.find(key);
+      assert(lhsIt != map.end() || rhsIt != rhs.map.end());
+
+      // If present in both, join.
+      if (lhsIt != map.end() && rhsIt != rhs.map.end()) {
+        result |= lhsIt->getSecond().join(rhsIt->getSecond());
+        continue;
+      }
+
+      // Copy from RHS if available only there.
+      if (lhsIt == map.end()) {
+        map.try_emplace(rhsIt->getFirst(), rhsIt->getSecond());
+        result = ChangeResult::Change;
+      }
+
+      // Do nothing if available only in LHS.
+    }
+    return result;
+  }
+
+  /// Map all keys to all values.
+  ChangeResult insert(const SetLattice<KeyT> &keysToUpdate,
+                      const SetLattice<ElementT> &values) {
+    if (keysToUpdate.isUnknown())
+      return markAllUnknown();
+
+    if (keysToUpdate.isUndefined())
+      return ChangeResult::NoChange;
+
+    return keysToUpdate.foreachElement(
+        [&](DistinctAttr key, typename SetLattice<KeyT>::State state) {
+          assert(state == SetLattice<KeyT>::State::Defined &&
+                 "unknown must have been handled above");
+          return joinPotentiallyMissing(key, values);
+        });
+  }
+
+  ChangeResult markAllUnknown() {
+    ChangeResult result = ChangeResult::NoChange;
+    for (auto &it : map)
+      result |= it.getSecond().join(SetLattice<ElementT>::getUnknown());
+    return result;
+  }
+
+  const SetLattice<ElementT> &lookup(KeyT key) const {
+    auto it = map.find(key);
+    if (it == map.end())
+      return SetLattice<ElementT>::getUndefined();
+    return it->getSecond();
+  }
+
+protected:
+  ChangeResult joinPotentiallyMissing(KeyT key,
+                                      const SetLattice<ElementT> &value) {
+    // Don't store explicitly undefined values in the mapping, keys absent from
+    // the mapping are treated as implicitly undefined.
+    if (value.isUndefined())
+      return ChangeResult::NoChange;
+
+    bool inserted;
+    decltype(map.begin()) iterator;
+    std::tie(iterator, inserted) = map.try_emplace(key, value);
+    if (!inserted)
+      return iterator->second.join(value);
+    return ChangeResult::Change;
+  }
+
+  /// Maps a key to a set of values. When a key is not present in this map, it
+  /// is considered to map to an uninitialized set.
+  DenseMap<KeyT, SetLattice<ElementT>> map;
+
+private:
+  Attribute serializeMapOfSetsNaive(MLIRContext *ctx) const {
+    SmallVector<Attribute> pointsToArray;
+
+    for (const auto &[srcClass, destClasses] : map) {
+      SmallVector<Attribute, 2> pair = {srcClass};
+      SmallVector<Attribute, 5> aliasClasses;
+      if (destClasses.isUnknown()) {
+        aliasClasses.push_back(StringAttr::get(ctx, "unknown"));
+      } else if (destClasses.isUndefined()) {
+        aliasClasses.push_back(StringAttr::get(ctx, "undefined"));
+      } else {
```

Comment:
Can we unify these string representations with what is used above (e.g., `"<unknown>"`) by putting the strings in some named constant?